server/public/model/report.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"net/http"
+	"time"
+
+	pUtils "github.com/mattermost/mattermost/server/public/utils"
+)
+
+const (
+	ReportDurationLast30Days    = "last_30_days"
+	ReportDurationPreviousMonth = "previous_month"
+	ReportDurationLast6Months   = "last_6_months"
+
+	ReportingMaxPageSize = 100
+)
+
+var (
+	UserReportSortColumns = []string{"CreateAt", "Username", "FirstName", "LastName", "Nickname", "Email", "Roles"}
+
+	ChannelReportSortColumns = []string{"ChannelName", "MemberCount", "PostCount", "TeamName", "ChannelManagement"}
+)
+
+type ReportingBaseOptions struct {
+	SortDesc            bool
+	PageSize            int
+	SortColumn          string
+	LastSortColumnValue string
+	DateRange           string
+	StartAt             int64
+	EndAt               int64
+}
+
+func (options *ReportingBaseOptions) PopulateDateRange(now time.Time) {
+	startAt := int64(0)
+	endAt := int64(0)
+
+	if options.DateRange == ReportDurationLast30Days {
+		startAt = now.AddDate(0, 0, -30).UnixMilli()
+	} else if options.DateRange == ReportDurationPreviousMonth {
+		startOfMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.Local)
+		startAt = startOfMonth.AddDate(0, -1, 0).UnixMilli()
+		endAt = startOfMonth.UnixMilli()
+	} else if options.DateRange == ReportDurationLast6Months {
+		startAt = now.AddDate(0, -6, -0).UnixMilli()
+	}
+
+	options.StartAt = startAt
+	options.EndAt = endAt
+}
+
+func (options *ReportingBaseOptions) IsValid() *AppError {
+	// Don't allow fetching more than 100 users at a time from the normal query endpoint
+	if options.PageSize <= 0 || options.PageSize > ReportingMaxPageSize {
```

Comment:
These checks aren't really a property of the model itself, they more exist to protect the API. I originally had them in the `app` module but I'm starting to think they just belong in the API itself. So I think we should move them over there.

Reply:
Makes sense. Done.